### DISCUSSION TITLE
How can I run detect.py on multiple Nodes - Multi-CPU Training

### DISCUSSION THREAD

**Discussion by x8n1620** [ultralytics/yolov5#7020] [Q&A]
Hey there,
I have a bunch of very potent Server with much unused CPU Resources outside of the business hours (14h a day).
I run yolov5 train.py on a single node with 40 CPU-Cores and 1TB RAM on my dataset, it takes more than 3 Nights to process.
I've many of these nodes, so I try to setup a Distributed Data Parallel like this https://github.com/ultralytics/yolov5/issues/475 with torch.distributed.launch.
I have only CPU-Power, no GPUs with CUDA inside of these nodes, and so train.py fails with errors

My setup:
Node 1:
python3 -m torch.distributed.launch --nproc_per_node=1 --nnodes=2 --node_rank=0 --master_addr="x.x.x.x" --master_port=24400 train.py --img 640 --batch 4 --epochs 2 --data ../sshfs/data/data.yaml --weights yolov5n.pt --cache --device cpu --project ../sshfs/out --name exp
Node 2:
python3 -m torch.distributed.launch --nproc_per_node=1 --nnodes=2 --node_rank=1 --master_addr="x.x.x.x" --master_port=24400 train.py --img 640 --batch 4 --epochs 2 --data ../sshfs/data/data.yaml --weights yolov5n.pt --cache --device cpu --project ../sshfs/out --name exp

../sshfs/ is shared Storage.

Is it possible tu run train.py trough torch.distributed.launch on multiple nodes with CPU-Usage?

Regards

**Comment by glenn-jocher** [comment]
@x8n1620 I'm sorry I don't understand, are you trying to training or detecting? Your question is about detecting but your code examples show training commands.

**Comment by glenn-jocher** [comment]
@x8n1620 also in general torch.distributed.run commands are for multi-GPU training. There won't be any benefits to running this with CPU training.

**Comment by x8n1620** [comment]
Oh, sorry for my confusion! My question is: how to train (!) on multiple Nodes with CPUs only.
Why is there no benefit torch.distributed to run  torch.distributed on multiple nodes? 80 CPU-Cores in two nodes are faster than 40 in one. And I have a lot of nodes with 40 CPU Cores each.
PyTorch describes distributed CPU training here:
https://pytorch.org/docs/stable/distributed.html

**Reply by glenn-jocher** [reply]
@x8n1620 hmm well maybe you are right and there are advantages. I haven't trained multi-node before though so I don't think I can help too much. It does seem from the docs though that the multi-node examples are geared towards multi-GPU.

**Comment by x8n1620** [comment]
In Line 60
LOCAL_RANK = int(os.getenv('LOCAL_RANK', -1))  # https://pytorch.org/docs/stable/elastic/run.html
followed of Line 526 yolov5 train.py ask in MultiMode, if multiple cuda capable devices are available.
It seems, it is not ready for multi node runs in CPU mode.
Is that right? And when yes: why?

**Reply by glenn-jocher** [reply]
@x8n1620 `LOCAL_RANK` environment variables are defined by `torch.distributed.run`, we don't control these.
